pyteal/ast/router.py

```diff
@@ -0,0 +1,583 @@
+                    f"got {handler.subroutine.argument_count()} args with {len(handler.subroutine.abi_args)} ABI args."
+                )
+
+            arg_type_specs = cast(
+                list[abi.TypeSpec], handler.subroutine.expected_arg_types
+            )
+            if handler.subroutine.argument_count() > METHOD_ARG_NUM_LIMIT:
+                last_arg_specs_grouped = arg_type_specs[METHOD_ARG_NUM_LIMIT - 1 :]
+                arg_type_specs = arg_type_specs[: METHOD_ARG_NUM_LIMIT - 1]
+                last_arg_spec = abi.TupleTypeSpec(*last_arg_specs_grouped)
+                arg_type_specs.append(last_arg_spec)
+
+            arg_abi_vars: list[abi.BaseType] = [
+                type_spec.new_instance() for type_spec in arg_type_specs
+            ]
+            decode_instructions: list[Expr] = [
+                arg_abi_vars[i].decode(Txn.application_args[i + 1])
+                for i in range(len(arg_type_specs))
+            ]
+
+            if handler.subroutine.argument_count() > METHOD_ARG_NUM_LIMIT:
+                tuple_arg_type_specs: list[abi.TypeSpec] = cast(
+                    list[abi.TypeSpec],
+                    handler.subroutine.expected_arg_types[METHOD_ARG_NUM_LIMIT - 1 :],
+                )
+                tuple_abi_args: list[abi.BaseType] = [
+                    t_arg_ts.new_instance() for t_arg_ts in tuple_arg_type_specs
+                ]
+                last_tuple_arg: abi.Tuple = cast(abi.Tuple, arg_abi_vars[-1])
+                de_tuple_instructions: list[Expr] = [
+                    last_tuple_arg[i].store_into(tuple_abi_args[i])
+                    for i in range(len(tuple_arg_type_specs))
+                ]
+                decode_instructions += de_tuple_instructions
+                arg_abi_vars = arg_abi_vars[:-1] + tuple_abi_args
+
+            # NOTE: does not have to have return, can be void method
+            if handler.type_of() == "void":
+                return Seq(
+                    *decode_instructions,
+                    cast(Expr, handler(*arg_abi_vars)),
+                    Approve(),
+                )
+            else:
+                output_temp: abi.BaseType = cast(
+                    OutputKwArgInfo, handler.output_kwarg_info
+                ).abi_type.new_instance()
+                subroutine_call: abi.ReturnedValue = cast(
+                    abi.ReturnedValue, handler(*arg_abi_vars)
+                )
+                return Seq(
+                    *decode_instructions,
+                    subroutine_call.store_into(output_temp),
+                    abi.MethodReturn(output_temp),
+                    Approve(),
+                )
+
+    def __add_bare_call(self, oc_actions: OnCompleteActions) -> None:
+        action_type = Expr | SubroutineFnWrapper | ABIReturnSubroutine
+        if oc_actions.is_empty():
+            raise TealInternalError("the OnCompleteActions is empty.")
```

Comment:
It should be ok for the contract to support no bare app calls

Reply:
👍